### PR TITLE
chore: refactor clients

### PR DESCRIPTION
**Pull request by ldez**
TLDR: rewrite all the internal clients to be more uniform, have better errors, and improve debugging. 

---

My journey on this PR started in December (yes December) when I was reviewing a PR about a new DNS provider.
I was thinking that the tests are really not homogeneous, and they are painful to write.
Then I started to think about something to simplify the way to write tests.
So I started but at this point, I was struck by the non-uniformity of the clients themselves.

I started to think about how to have something more uniform without using or creating a framework because each provider has its own small variation.
I rewrite all clients multiple times, and I tried multiple approaches for authentication, error management, etc.
I rewrote each client "layer" by "layer" to try to have something more uniform, then I increase the "duplication" on purpose.
I think it's easier to contribute if the "shape" of the clients is homogeneous even if they contain small variations.

Also, during this review/rewrite, I saw that the errors are never the same and didn't provide the expected information.
Then I created some shared error types.

In the end, I didn't rewrite the tests in this PR, because the PR is already too large, so the tests will be rewritten in another PR, but I added missing tests to help during the refactor.

This PR is large, and I spent a lot of time on it, so @dmke, I hope the review of the PR will be less complex than the analysis of all API clients.

I hope this PR will increase the maintainability, and facilitate the creation and review of new DNS providers. 

Now I will be available for reviewing the other PRs (even if I already prepared 3 new branches :smile:)

### CODE REVIEW

**Comment by dmke** [comment]
That's... huge. I've glanced over it, but haven't had the time to perform a proper review (as mentioned elsewhere, I'm quite busy until at least this weekend).

So far, I like what I see :)

**Comment by dmke** [comment]
Sorry for the long pause, I'll try to review this in the coming days.

**Review by dmke — APPROVED** [review]
Uff. This was a long one :-)

From a technical point of view, I don't have anything to remark. This is a straight forward preparation for the challenge provider methods to receive a `context.Context` in the (near?) future.

I've tested a few updated providers, and didn't note any faults (though the changes are rather minimal):

- hetzner
- cloudflare (no functional changes)
- googledomains (no changes at all)
- rfc2136 (no functional changes)

(I do also have a domain at Netcup, but their primary DNS server seems slow to update; the default timeout of 2min might be too short. I could observe the correct token in the management UI though).

Speaking of timeouts, one tidbit I've noticed are different names for the TTL values:

```console
$ egrep -r 'const .*TTL' . | awk -F: '{ print $1 "|" $2 }' | sort | column -t -s '|'
./providers/dns/arvancloud/arvancloud.go    const minTTL = 600
./providers/dns/autodns/autodns.go          const defaultTTL int = 600
./providers/dns/brandit/brandit.go          const defaultTTL = 600
./providers/dns/bunny/bunny.go              const minTTL = 60
./providers/dns/checkdomain/checkdomain.go  const defaultTTL = 300
./providers/dns/desec/desec.go              const defaultTTL int = 3600
./providers/dns/gandi/gandi.go              const minTTL = 300
./providers/dns/gandiv5/gandiv5.go          const minTTL = 300
./providers/dns/godaddy/godaddy.go          const minTTL = 600
./providers/dns/hetzner/hetzner.go          const minTTL = 60
./providers/dns/ionos/ionos.go              const minTTL = 300
./providers/dns/loopia/loopia.go            const minTTL = 300
./providers/dns/luadns/luadns.go            const minTTL = 300
./providers/dns/namedotcom/namedotcom.go    const minTTL = 300
./providers/dns/nicmanager/nicmanager.go    const minTTL = 900
./providers/dns/otc/otc.go                  const minTTL = 300
./providers/dns/porkbun/porkbun.go          const minTTL = 300
./providers/dns/selectel/selectel.go        const minTTL = 60
./providers/dns/servercow/servercow.go      const defaultTTL = 120
./providers/dns/variomedia/variomedia.go    const defaultTTL = 300
./providers/dns/vkcloud/vkcloud.go          const defaultTTL = 60
./providers/dns/vscale/vscale.go            const minTTL = 60
./providers/dns/websupport/websupport.go    const defaultTTL = 600
./providers/dns/wedos/wedos.go              const minTTL = 5 * 60 // 5 minutes
./providers/dns/yandexcloud/yandexcloud.go  const defaultTTL = 60
./providers/dns/yandex/yandex.go            const defaultTTL = 21600
```

With this PR being mostly a cleanup PR, do you want to normalize `minTTL` to `defaultTTL`?

**Comment by ldez** [comment]
> With this PR being mostly a cleanup PR, do you want to normalize `minTTL` to `defaultTTL`?

It depends on the provider, in some cases the provider has a real minimum TTL.
I will check that.

**Comment by ldez** [comment]
I removed `defaultTTL` constants.

**Review by dmke — APPROVED** [review]
